nowcasting_dataset/manager/manager_live.py

```diff
@@ -0,0 +1,236 @@
+
+        logger.info(
+            f"{SPATIAL_AND_TEMPORAL_LOCATIONS_OF_EACH_EXAMPLE_FILENAME} does not exist so"
+            " will create..."
+        )
+
+        split_name = "live"
+        datetimes_for_split = [t0_datetime]
+
+        path_for_csv = self.config.output_data.filepath / split_name
+        # TODO make dynamic
+        n_batches_requested = int(np.ceil(N_GSPS / self.config.process.batch_size))
+
+        n_examples = n_batches_requested * self.config.process.batch_size
+        logger.debug(
+            f"Creating {n_batches_requested:,d} batches x {self.config.process.batch_size:,d}"
+            f" examples per batch = {n_examples:,d} examples for {split_name}."
+        )
+
+        df_of_locations = self.sample_spatial_and_temporal_locations_for_examples(
+            t0_datetime=datetimes_for_split[0],
+        )
+        output_filename = self._filename_of_locations_csv_file(split_name="live")
+        logger.info(f"Making {path_for_csv} if it does not exist.")
+        nd_fs_utils.makedirs(path_for_csv, exist_ok=True)
+        logger.debug(f"Writing {output_filename}")
+        df_of_locations.to_csv(output_filename)
+
+    def sample_spatial_and_temporal_locations_for_examples(
+        self, t0_datetime: datetime
+    ) -> pd.DataFrame:
+        """
+        Computes the geospatial and temporal locations for each training example.
+
+        The first data_source in this DataSourceList defines the geospatial locations of
+        each example.
+
+        Args:
+            t0_datetimes: All available t0 datetimes.  Can be computed with
+                `DataSourceList.get_t0_datetimes_across_all_data_sources()`
+            n_examples: The number of examples requested.
+            get_all_locations: optional to return all locations for each t0_datetime
+
+        Returns:
+            Each row of each the DataFrame specifies the position of each example, using
+            columns: 't0_datetime_UTC', 'x_center_OSGB', 'y_center_OSGB'.
+        """
+
+        # note that the returned 'shuffled_t0_datetimes'
+        # has duplicate datetimes for each location
+        (
+            shuffled_t0_datetimes,
+            x_locations,
+            y_locations,
+        ) = self.data_source_which_defines_geospatial_locations.get_all_locations(
+            t0_datetimes_utc=pd.DatetimeIndex([t0_datetime])
+        )
+        shuffled_t0_datetimes = list(shuffled_t0_datetimes)
+
+        # find out the number of examples in the last batch,
+        # we maybe need to duplicate the last example into order to get a full batch
+        n_examples_last_batch = len(shuffled_t0_datetimes) % self.config.process.batch_size
+        # Note 0 means the examples fit into the batches
+
+        if n_examples_last_batch != 0:
+            # add extra examples to make sure the number of examples fits into batches.
+            # There could be a different way to do this,
+            # but this keeps it pretty simple for the moment
```

Comment:
Potentially we could reduce the batch size to a multiple of the number of GSPs? E.g. 338 is divisible by 13 or 26, so could use a batch of 26 instead?

Reply:
That is true, although I don't think we should pin ourselves to a batch size that divides into 338.